File: day12.c
#include <stdio.h>
#define MAXCC 1000

int main(void){
    char G[150][150];
    FILE *f=fopen("data/input12.txt","r");
    int N=0;
    while(fgets(G[N],150,f) && G[N][0]!='\n'){N++;}
    int CC[150][150]={0};
    int numcc=1;
    void dfs(int x,int y){
        if(CC[y][x]>0)return;
        CC[y][x]=numcc;
        if(x>0 && G[y][x-1]==G[y][x])dfs(x-1,y);
        if(x<N-1 && G[y][x+1]==G[y][x])dfs(x+1,y);
        if(y>0 && G[y-1][x]==G[y][x])dfs(x,y-1);
        if(y<N-1 && G[y+1][x]==G[y][x])dfs(x,y+1);
    }
    for(int y=0; y<N; y++){
        for(int x=0; x<N; x++){
            if(CC[y][x]==0){dfs(x,y);numcc++;}
        }
    }
    int get(int x,int y){
        if(x<0||x>=N||y<0||y>=N)return -1;
        return CC[y][x];
    }
    int area[MAXCC]={0},perim[MAXCC]={0},corners[MAXCC]={0};
    for(int y=0; y<N; y++){
        for(int x=0; x<N; x++){
            for(int dy=-1; dy<=1; dy+=2){
                for(int dx=-1; dx<=1; dx+=2){
                    int a=get(x,y),b=get(x+dx,y),c=get(x,y+dy),d=get(x+dx,y+dy);
                    if((a!=b && a!=c)||(a==b && a==c && a!=d)){corners[a]++;}
                }
            }
            int k=4;
            int a=get(x,y),b=get(x-1,y),c=get(x,y-1);
            if(a==b)k-=2;
            if(a==c)k-=2;
            area[a]++; perim[a]+=k;
        }
    }
    int part1=0,part2=0;
    for(int i=0;i<MAXCC;i++){part1+=area[i]*perim[i]; part2+=area[i]*corners[i];}
    printf("%d %d\n",part1,part2);
    return 0;
}

File: day12.lua
local part1,part2=0,0
local G={}
local H,W=0,0
for x in io.lines("data/input12.txt") do
    if #x>0 then
        table.insert(G,x);H=H+1
    end
end
W=#G[1]
local numcc=0
local CC={}
function dfs(x,y)
    local code=W*y+x
    if CC[code] then return end
    CC[code]=numcc
    local h=string.sub(G[y],x,x)
    if x>1 and h==string.sub(G[y],x-1,x-1) then dfs(x-1,y) end
    if x<W and h==string.sub(G[y],x+1,x+1) then dfs(x+1,y) end
    if y>1 and h==string.sub(G[y-1],x,x) then dfs(x,y-1) end
    if y<H and h==string.sub(G[y+1],x,x) then dfs(x,y+1) end
end
for y=1,H do
    for x=1,W do
        if not CC[y*W+x] then dfs(x,y); numcc=numcc+1; end
    end
end

function get(x,y) local code=y*W+x; return CC[code] end
local area,perim,corners={},{},{}
for y=1,H do
    for x=1,W do
        for _,dx in ipairs({-1,1}) do
            for _,dy in ipairs({-1,1}) do
                local a,b,c,d=get(x,y),get(x+dx,y),get(x,y+dy),get(x+dx,y+dy)
                if (a~=b and a~=c) or (a==b and a==c and a~=d)then
                    corners[a]=corners[a] and corners[a]+1 or 1
                end
            end
        end
        local a,b,c=get(x,y),get(x-1,y),get(x,y-1)
        local k=4
        if a==b then k=k-2 end
        if a==c then k=k-2 end
        area[a]=area[a] and area[a]+1 or 1
        perim[a]=perim[a] and perim[a]+k or k
    end
end

for k,v in pairs(area) do
    part1=part1+area[k]*perim[k]
    part2=part2+area[k]*corners[k]
end
print(part1,part2)